faucet/src/faucet.rs

```diff
@@ -625,44 +625,54 @@ async fn break_up_records(state: FaucetState, mut wakeup: mpsc::Receiver<()>) {
                 break;
             }
 
-            // Find a record which can be broken down into two smaller `grant_size` records.
-            let record = match records
+            let largest_record = match records
                 .into_iter()
-                .find(|record| record.amount() > state.grant_size * 2u64)
+                .max_by(|x, y| x.amount().cmp(&y.amount()))
             {
                 Some(record) => record,
                 None => {
-                    // There are no big records to break up, so there's nothing for us to do. Exit
-                    // the inner loop and wait for a notification that the record distribution has
-                    // changed.
+                    tracing::warn!("No spendable records");
                     break;
                 }
             };
 
+            if largest_record.amount() < state.grant_size * 2u64 {
+                // There are no big records to break up, so there's nothing for us to do. Exit
+                // the inner loop and wait for a notification that the record distribution has
+                // changed.
+                tracing::warn!("No large records to break up");
+                break;
+            };
+
+            let split_amount = largest_record.amount() / 2;
+            let change_amount = largest_record.amount() - split_amount;
+
             tracing::info!(
                 "breaking up a record of size {} into records of size {} and {}",
-                record.amount(),
-                state.grant_size,
-                record.amount() - state.grant_size
+                largest_record.amount(),
+                split_amount,
+                change_amount,
             );
 
             // There is not yet an interface for transferring a specific record, so we just have to
             // specify the appropriate amounts and trust that Seahorse will use the largest record
-            // available (it should). Just to be extra safe, we specify the larger of the two
-            // amounts -- record.amount() - state.grant_size -- as the output amount, which should
-            // create a change record of size `state.grant_size`. This makes it impossible for
-            // Seahorse to choose a record of exactly `state.grant_size` with no change, which would
-            // prevent this loop from making progress.
+            // available (it should). We specify two outputs so that if an existing record with
+            // `change_amount` exists it won't be used "as is", which would prevent this loop
+            // from making progress.
             let receipt = match wallet
                 .transfer(
                     None,
                     &AssetCode::native(),
-                    &[(address.clone(), record.amount() - state.grant_size)],
+                    &[
+                        (address.clone(), change_amount),
+                        (address.clone(), split_amount),
```

Comment:
For reference, here are the test logs *if the second recipient entry is removed* (plus debug printing of all spendable records).

When we get to 3 records the large one `85070591730234615865843651857942052863` should be selected but is not.

```
running 1 test
Unpacking universal parameters... done in 9761 ms
Sent funding tx to deployer
xfr: 987 bytes
mint: 971 bytes
freeze: 979 bytes
2022-06-23T14:07:31.416098Z  INFO surf::middleware::logger::native: sending request    
2022-06-23T14:07:31.416650Z  WARN surf::middleware::logger::native: request completed    
Unpacking universal parameters... done in 9957 ms
2022-06-23T14:07:41.387293Z  INFO surf::middleware::logger::native: sending request    
2022-06-23T14:07:41.387680Z  WARN surf::middleware::logger::native: request completed    
test faucet::test::test_faucet_transfer has been running for over 60 seconds
2022-06-23T14:07:45.329152Z  INFO tide::server: Server listening on http://0.0.0.0:15600    
2022-06-23T14:07:45.847669Z  INFO faucet: Wallet balance before init: 170141183460469231731687303715884105727
2022-06-23T14:07:45.847698Z  INFO faucet: have 1 records, waiting for 4 more
spendable RecordAmount(Amount(170141183460469231731687303715884105727))
2022-06-23T14:07:45.847724Z  INFO faucet: breaking up a record of size 170141183460469231731687303715884105727 into records of size 85070591730234615865843651857942052863 and 85070591730234615865843651857942052864
spendable RecordAmount(Amount(85070591730234615865843651857942052863))
spendable RecordAmount(Amount(85070591730234615865843651857942052864))
2022-06-23T14:07:49.360818Z  INFO faucet: breaking up a record of size 85070591730234615865843651857942052864 into records of size 42535295865117307932921825928971026432 and 42535295865117307932921825928971026432
2022-06-23T14:07:53.395379Z  INFO faucet: have 3 records, waiting for 2 more
spendable RecordAmount(Amount(85070591730234615865843651857942052863))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
2022-06-23T14:07:53.395436Z  INFO faucet: breaking up a record of size 85070591730234615865843651857942052863 into records of size 42535295865117307932921825928971026431 and 42535295865117307932921825928971026432
2022-06-23T14:07:57.429710Z  INFO faucet: have 3 records, waiting for 2 more
spendable RecordAmount(Amount(85070591730234615865843651857942052863))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
2022-06-23T14:07:57.429756Z  INFO faucet: breaking up a record of size 85070591730234615865843651857942052863 into records of size 42535295865117307932921825928971026431 and 42535295865117307932921825928971026432
2022-06-23T14:08:01.410172Z  INFO tide::log::middleware: <-- Request received    
2022-06-23T14:08:01.410231Z  INFO tide::log::middleware: --> Response sent    
2022-06-23T14:08:01.464903Z  INFO faucet: have 3 records, waiting for 2 more
spendable RecordAmount(Amount(85070591730234615865843651857942052863))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
2022-06-23T14:08:01.464939Z  INFO faucet: breaking up a record of size 85070591730234615865843651857942052863 into records of size 42535295865117307932921825928971026431 and 42535295865117307932921825928971026432
2022-06-23T14:08:05.500190Z  INFO faucet: have 3 records, waiting for 2 more
spendable RecordAmount(Amount(85070591730234615865843651857942052863))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
spendable RecordAmount(Amount(42535295865117307932921825928971026432))
2022-06-23T14:08:05.500235Z  INFO faucet: breaking up a record of size 85070591730234615865843651857942052863 into records of size 42535295865117307932921825928971026431 and 42535295865117307932921825928971026432
```

Reply:
Oh, yeah, you are running into the thing where it will allocate a smaller record if it can do so without change: https://github.com/EspressoSystems/seahorse/blob/00853e470e577535663b8293af772fc79d3cdb4f/src/txn_builder.rs#L1691. Using both outputs explicitly is a good solution; since we always split up the largest record, it forces Seahorse to use this record.